pandas/core/reshape/reshape.py

```diff
@@ -1072,6 +1072,223 @@ def get_empty_frame(data) -> DataFrame:
+        )
+
+    if to_series:
+        return _from_dummies_1d(data, dropped_first)
+
+    data_to_decode: DataFrame
+    if columns is None:
+        columns = data.columns
+    elif not is_list_like(columns):
+        raise TypeError("Argument for parameter 'columns' must be list-like")
+    # index data with a list of all columns that are dummies
+    cat_columns = []
+    non_cat_columns = []
+    for col in columns:
+        if any(ps in col for ps in prefix_sep):
+            cat_columns.append(col)
+        else:
+            non_cat_columns.append(col)
+    data_to_decode = data[cat_columns].astype("boolean")
+    non_cat_data = data[non_cat_columns]
+
+    # get separator for each prefix and lists to slice data for each prefix
+    if isinstance(prefix_sep, dict):
+        variables_slice: dict[str, list] = {prefix: [] for prefix in prefix_sep}
+        for col in data_to_decode.columns:
+            for prefix in prefix_sep:
+                if prefix in col:
+                    variables_slice[prefix].append(col)
+    else:
+        sep_for_prefix = {}
+        variables_slice = {}
+        for col in data_to_decode.columns:
+            ps = [ps for ps in prefix_sep if ps in col][0]
+            prefix = col.split(ps)[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        prefix_sep = sep_for_prefix
+
+    # validate number of dropped_first
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})."
+            )
+            raise ValueError(len_msg)
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif isinstance(dropped_first, list):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
+        else:
+            dropped_first = dict(
+                zip(variables_slice, [dropped_first] * len(variables_slice))
+            )
+
+    cat_data = {}
+    for prefix, prefix_slice in variables_slice.items():
+        cats = [col[len(prefix + prefix_sep[prefix]) :] for col in prefix_slice]
+        assigned = data_to_decode[prefix_slice].sum(axis=1)
+        if any(assigned > 1):
+            raise ValueError(
+                f"Dummy DataFrame contains multi-assignment(s) for prefix: "
+                f"'{prefix}'; First instance in row: {assigned.argmax()}."
+            )
+        elif any(assigned == 0):
+            if isinstance(dropped_first, dict):
+                cats.append(dropped_first[prefix])
```

Comment:
Not as elegant as `if dropped_first:` but required by mypy - it could be any of the passed types from a mypy point of view.

Reply:
You could `assert isinstance(dropped_first, dict)`.  You might want to use `Mapping` which would cover a wider class.